stl/inc/mdspan

```diff
@@ -0,0 +1,1464 @@
+        array<rank_type, _Rank + 1> _Result{};
+        rank_type _Counter = 0;
+        for (rank_type _Idx = 0; _Idx < _Rank; ++_Idx) {
+            _Result[_Idx] = _Counter;
+            if (_Static_extents[_Idx] == dynamic_extent) {
+                ++_Counter;
+            }
+        }
+        _Result[_Rank] = _Counter;
+        return _Result;
+    }
+
+    static constexpr array<rank_type, _Rank + 1> _Dynamic_indices = _Make_dynamic_indices();
+
+    _NODISCARD static consteval auto _Make_dynamic_indices_inv() noexcept {
+        array<rank_type, _Rank_dynamic> _Result{};
+        rank_type _Counter = 0;
+        for (rank_type _Idx = 0; _Idx < _Rank; ++_Idx) {
+            if (_Static_extents[_Idx] == dynamic_extent) {
+                _Analysis_assume_(_Counter < _Rank_dynamic); // TRANSITION, DevCom-923103
+                _Result[_Counter] = _Idx;
+                ++_Counter;
+            }
+        }
+        return _Result;
+    }
+
+    static constexpr array<rank_type, _Rank_dynamic> _Dynamic_indices_inv = _Make_dynamic_indices_inv();
+
+    using _Base = _Maybe_empty_array<_IndexType, _Rank_dynamic>;
+
+    template <class _OtherIndexType, size_t... _OtherExtents, size_t... _Indices>
+        requires (sizeof...(_OtherExtents) == _Rank)
+              && ((_OtherExtents == dynamic_extent || _Extents == dynamic_extent || _OtherExtents == _Extents) && ...)
+    constexpr explicit extents(
+        const extents<_OtherIndexType, _OtherExtents...>& _Other, index_sequence<_Indices...>) noexcept
+        : _Base{static_cast<index_type>(_Other.extent(_Dynamic_indices_inv[_Indices]))...} {
+#if _CONTAINER_DEBUG_LEVEL > 0
+        if constexpr (rank() > 0) {
+            for (rank_type _Idx = 0; _Idx < _Rank; ++_Idx) {
+                if constexpr (rank() != rank_dynamic()) {
+                    _STL_VERIFY(_Static_extents[_Idx] == dynamic_extent
+                                    || _STD cmp_equal(_Static_extents[_Idx], _Other.extent(_Idx)),
+                        "Value of other.extent(r) must be equal to extent(r) for each r for which extent(r) is a "
+                        "static extent (N4950 [mdspan.extents.cons]/2.1)");
+                }
+                _STL_VERIFY(_STD in_range<index_type>(_Other.extent(_Idx)),
+                    "Value of other.extent(r) must be representable as a value of type index_type for every rank index "
+                    "r (N4950 [mdspan.extents.cons]/2.2)");
+            }
+        }
+#endif // _CONTAINER_DEBUG_LEVEL > 0
+    }
+
+    struct _Construct_from_tuple {
+        explicit _Construct_from_tuple() = default;
+    };
+
+    template <class _ExtsTuple, size_t... _Indices>
+        requires (tuple_size_v<_ExtsTuple> == _Rank_dynamic)
+    constexpr explicit extents(_Construct_from_tuple, _ExtsTuple _Tpl, index_sequence<_Indices...>) noexcept
+        : _Base{static_cast<index_type>(_STD move(_STD get<_Indices>(_Tpl)))...} {}

```

Comment:
I find this is specified by the standard; just curious, why does the standard say the elements should get `move`d before casting to an integer type?

Reply:
My understanding is that they're allowed to be class types that are convertible to integer types; the test suite has a few "weird indices" with such conversions. (And indeed they caught an over-simplification that I had briefly committed while working on my feedback changes, where I thought I could take a shortcut that was actually disallowed.)

Reply:
Yeah, generally constructors and other access operators in mdspan and friends are meant to allow to take things convertible to index_type (which itself however must be an integral). The primary intent is to allow for strongly typed indices. For example a bunch of our codes distinguish between things like edge and vertice indices. For most of those the move wouldn't really matter, but at that point LWG said: ok lets do this the right away even if the "indices" become even weirder.